clippy_lints/src/manual_clamp.rs

```diff
@@ -0,0 +1,720 @@
+            && let Some(first_bin) = try_binary_op(first_cond)
+            && let Some(second_bin) = try_binary_op(second_cond)
+            && let Some(input_min_max) = is_clamp_meta_pattern(
+                cx,
+                &first_bin,
+                &second_bin,
+                maybe_min_max_first,
+                maybe_min_max_second,
+                None
+            )
+            {
+                Some(ClampSuggestion {
+                    params: InputMinMax {
+                        input: init,
+                        min: input_min_max.min,
+                        max: input_min_max.max
+                    },
+                    add_semicolon: true,
+                    span: init.span.to(second_expr.span)
+                })
+            } else {
+                None
+            }
+                });
+        Some(iter)
+    } else {
+        None
+    };
+    ret.into_iter().flatten().collect()
+}
+
+fn simplify_kind<'tcx>(expr: &'tcx Expr<'tcx>) -> &'tcx Expr<'tcx> {
+    match &expr.kind {
+        ExprKind::Block(
+            Block {
+                stmts: &[],
+                expr: Some(e),
+                targeted_by_break: false,
+                rules: BlockCheckMode::DefaultBlock,
+                ..
+            },
+            _,
+        )
+        | ExprKind::DropTemps(e) => simplify_kind(e),
+        _ => expr,
+    }
+}
+
+fn is_clampable<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    ty.is_floating_point() || get_trait_def_id(cx, &paths::ORD).map_or(false, |id| implements_trait(cx, ty, id, &[]))
+}
+
+/// `ExprKind::Binary` but more narrowly typed
+struct BinaryOp<'tcx> {
+    op: BinOpKind,
+    left: &'tcx Expr<'tcx>,
+    right: &'tcx Expr<'tcx>,
+}
+
+fn try_binary_op(e: ExprRef<'_>) -> Option<BinaryOp<'_>> {
+    match &e.kind {
+        ExprKind::Binary(op, left, right) => Some(BinaryOp {
+            op: op.node,
+            left,
+            right,
+        }),
+        _ => None,
+    }
+}
+
+/// The clamp meta pattern is a pattern shared between many (but not all) patterns.
+/// In summary, this pattern consists of two if statements that meet many criteria,
+/// - binary operators that are one of [`>`, `<`, `>=`, `<=`].
+/// - Both binary statements must have a shared argument
+///     - Which can appear on the left or right side of either statement
+///     - The binary operators must define a finite range for the shared argument. To put this in
+///       the terms of Rust `std` library, the following ranges are acceptable
+///         - `Range`
+///         - `RangeInclusive`
+///       And all other range types are not accepted. For the purposes of `clamp` it's irrelevant
+///       whether the range is inclusive or not, the output is the same.
+/// - The result of each if statement must be equal to the argument unique to that if statement. The
+///   result can not be the shared argument in either case.
+fn is_clamp_meta_pattern<'tcx>(
```

Comment:
I think there's still a bit extra we can do here, if you have an inner function that checks only one variant, you could swap the args around before passing it to that. If you add a `BinaryOp::flip` function that swaps `.left` and `.right`, and the `.op` kind you'd be able to cover them all in a way that's easier to see

Reply:
That's a cool idea!! Implemented in fb16348